### PR TITLE
Feature/log rotate

### PR DESCRIPTION
**Pull request by kn-ms**
## Bug / Requirement Description
No Log rotation support in LogMatcher

## Solution description
Added Log rotation support. User can pass a glob instead of a logfile. Logfiles matching the glob ordered by mtime.

Some simpler helper class added as well, during the gradual refactoring. RotatingLogStream can be extended with different log rotation strategies.

## Checklist:
- [X] Test
- [ ] Example (both test_plan.py and .rst)
- [X] Documentation (API)
- [X] News fragment present for release notes
- [X] MS info leakage check
- [ ] For new driver: driver index page
- [ ] For new assertion: ui/pdf/std renderers, documentation
- [ ] For new cmdline arg: documentation

### CODE REVIEW

File: testplan/common/utils/logfile.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+from itertools import dropwhile
+from os import PathLike, SEEK_END
+from typing import Union, Optional, Generic, TypeVar, List
+
+
+class LogPosition:
+    pass
+
+
+T = TypeVar("T", str, bytes)
```
Comment (yuxuan-ms) [inline testplan/common/utils/logfile.py:15]: AnyStr

File: testplan/common/utils/logfile.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+        return self.file.readline()
+
+    @property
+    def position(self) -> FileLogPosition:
+        return self.FileLogPosition(position=self.file.tell())
+
+    def compare(
+        self, position1: FileLogPosition, position2: FileLogPosition
+    ) -> int:
+        return position1.position - position2.position
```
Comment (yuxuan-ms) [inline testplan/common/utils/logfile.py:89]: why not use `__sub__`?

File: testplan/common/utils/match.py [hunk 3]
```diff
@@ -132,14 +219,9 @@ def match(
...
-        # As a convenience, we create the compiled regex if a string was
-        # passed.
-        if not hasattr(regex, "match"):
-            regex = re.compile(regex)
-        if isinstance(regex.pattern, str):
-            read_mode = "r"
+        regex = self._prepare_regexp(regex)
 
-        with open(self.log_path, read_mode) as log:
+        with closing(self.log_stream) as log:
```
Comment (Pyifan) [inline testplan/common/utils/match.py:224]: Is it necessary to open-close the log_stream every time we match? I see the __del__ already do close, is that good enough?

File: testplan/common/utils/match.py [hunk 4]
```diff
@@ -155,7 +237,7 @@ def match(
                 else:
                     break
 
-            self.position = log.tell()
+            self.position = self.log_stream.position
```
Comment (Pyifan) [inline testplan/common/utils/match.py:240]: Also do we need to maintain the current position at LogMatch level now that the log_stream object also keeps record of current position. Maybe it's good enough to just keep a few marks at LogMatcher level? and delegate to log_stream to start from current position.

File: testplan/common/utils/match.py (hunk 3 above)
Comment (kn-ms) [reply testplan/common/utils/match.py:224]: That was the original behaviour. Having it in the del is just for a last resort, unbfortunately there are no guaranties that the logmatcher is already garbage collected when the drivers/runtime clean off logfiles in some of the cases. I wanted to be on the safe side here.

File: testplan/common/utils/match.py (hunk 4 above)
Comment (kn-ms) [reply testplan/common/utils/match.py:240]: log matcher keeps the position for continuous read, however if the stream is closed and reopened it is not maintained, taht I want to keep for now, see above.

File: testplan/common/utils/logfile.py [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+        position: int = 0
+
+    def __init__(self, path: Union[PathLike, str], binary: bool) -> None:
+        self._path = path
+        self._binary = binary
+        self._file: Optional[TextIOWrapper] = None
+
+    @property
+    def file(self):
+        mode = "rb" if self._binary else "rt"
```
[the remark this answers is not in the document]
Comment (kn-ms) [reply testplan/common/utils/logfile.py:67]: I went for full inheritence in the last update

File: testplan/common/utils/logfile.py (hunk 2 above)
Comment (kn-ms) [reply testplan/common/utils/logfile.py:92]: ordering in LogPosition could not be interpretted without the actual stream (like for rotated logfiles), I do not want the position have a refference to the stream. In the simple single file case of course they are comparable without the stream, but for the generic api the compare function should be a member of the stream,

**Review by Pyifan — APPROVED** [review]

**Pull request merged**